src/planning.py

```diff
@@ -417,6 +418,79 @@ def _run_low_level_search(task: Task, option_model: _OptionModelBase,
+    the option sequence successfully executed to achieve the goal and
+    generated a LowLevelTrajectory. Otherwise, it returns and empty list
+    and False.
+    """
+    start_time = time.time()
+    cur_idx = 0
+    # The number of actions taken by each option in the plan. This is to
+    # make sure that we do not exceed the task horizon.
+    num_actions_per_option = [0 for _ in plan]
+    traj: List[State] = [task.init] + [DefaultState for _ in plan]
+    actions: List[Action] = [Action(np.array([0.0])) for _ in plan]
+    while cur_idx < len(plan):
+        if time.time() - start_time > timeout:
+            return LowLevelTrajectory([], [], False), False
+        state = traj[cur_idx]
+        option = plan[cur_idx]
+        cur_idx += 1
+        if option.initiable(state):
+            try:
+                next_state, num_actions = \
+                    option_model.get_next_state_and_num_actions(state, option)
+                print('Success')
+            except EnvironmentFailure as e:
+                can_continue_on = False
+                return LowLevelTrajectory([], [], False), False
+            else:  # an EnvironmentFailure was not raised
+                num_actions_per_option[cur_idx - 1] = num_actions
+                traj[cur_idx] = next_state
+                action_option = ParameterizedOption(
+                    option.parent.name, option.parent.types,
+                    option.parent.params_space,
+                    lambda s, m, o, p: Action(np.array([0.0])),
+                    lambda s, m, o, p: False,
+                    lambda s, m, o, p: True).ground(option.objects,
+                                                    option.params)
+                action_option.name = option.name
+                action_option.memory = option.memory
+                actions[cur_idx - 1].set_option(action_option)
+                # Check if we have exceeded the horizon.
+                if np.sum(num_actions_per_option[:cur_idx]) > max_horizon:
+                    print(f"Max horizon {max_horizon} exceeded:",
+                          np.sum(num_actions_per_option[:cur_idx]), "steps")
+                    can_continue_on = False
+                else:
+                    # Since we're not checking the expected_atoms, we need to
+                    # explicitly check if the goal is achieved.
+                    can_continue_on = True
+                    if cur_idx == len(plan):
+                        if task.goal_holds(traj[cur_idx]):
+                            return LowLevelTrajectory(
+                                traj, actions), True  # success!
+                        can_continue_on = False
+                        return LowLevelTrajectory([], [], False), False
+        else:
+            # The option is not initiable.
+            can_continue_on = False
+        if not can_continue_on:  # we got stuck, time to return False!
+            return LowLevelTrajectory([], [], False), False
```

Comment:
oh, so we're not even doing backtracking, we're actually just returning failure here? in that case, we should be able to drastically simplify this function right? like we don't need `cur_idx` and stuff, all we need to do is loop `for option in plan`, and for each one, call the option model to get the next state, and build up the `traj` and `actions` as we go. if any option is not initiable, or if horizon is reached, or if the goal isn't achieved at the end, we can return failure

Reply:
I don't think we have such a util yet. Did you mean to say "one action per option"?

Reply:
(Like why would there just be one action per option in general?)

Reply:
(they use this in the context of generating demonstrations for behavior, since the actions in the demos don't really matter / aren't used)

Reply:
ah right got it. so yeah I think we definitely would not have already implemented that util

Reply:
I made it simpler (no backchaining). The reason it is still in planning.py is because I got a bad circular-imports error when I tried to move it to utils @ronuchit .

Reply:
ah lol ok i've been thru that before